### PR TITLE
WIP: Updating/Adding to analysis functions for ROI sidebar

### PR DESCRIPTION
**Pull request by SaOgaz**
This addresses #461. @robelgeda will be adding the GUI display for the new functions.

### CODE REVIEW

**Comment by brechmos-stsci** [comment]
@robelgeda, this is the PR that will fix https://github.com/spacetelescope/specviz/issues/474, right? (or is it going to be a separate PR?

**Comment by robelgeda** [comment]
@brechmos-stsci Yes it will. My part is coming up.

**Comment by robelgeda** [comment]
@SaOgaz : I made a PR at [here](https://github.com/SaOgaz/specviz/pull/1). Once you merge it, it should fix #474 : 
screen shot 2018-10-31 at 12 46 29 pm

**Comment by SaOgaz** [comment]
Handing rest of this PR off to Robel.  Can help with adding tests when we get there. The two other things need that need to be done still are:

- Fixing the updates to the sidebar units when the spectrum display units are changed.
- It looks like we have the stats_widget code repeated in another file: widgets/statistics.py

**Comment by nluetzge** [comment]
I've been playing around with this and I still get a lot of crashes (with the COS_FUV spectrum). Which spectrum are you using to test it? As @gderosa2004  mentioned in #462 we should **not** use the combined spectrum anymore (since this was manipulated and does not reflect the format properly), but the COS_FUV.fits instead.

I'm having trouble to reproduce the crashes. But here are two error messages that I got. It happened when I moved the edges of the ROI or when I moved the ROI.. but not always.

```
Traceback (most recent call last):
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/statistics/statistics_widget.py", line 318, in update_statistics
    self.stats = compute_stats(spec)
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/statistics/statistics_widget.py", line 76, in compute_stats
    'fwhm': fwhm(spectrum),
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/width.py", line 105, in fwhm
    return computation_wrapper(_compute_fwhm, spectrum, regions)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/utils.py", line 19, in computation_wrapper
    return func(spectrum, **kwargs)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/width.py", line 173, in _compute_fwhm
    return _compute_single_fwhm(flux, spectral_axis)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/width.py", line 151, in _compute_single_fwhm
    l_idx = np.where(left == True)[0][-1]
IndexError: index -1 is out of bounds for axis 0 with size 0
Abort trap: 6
```

```
Traceback (most recent call last):
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/statistics/statistics_widget.py", line 318, in update_statistics
    self.stats = compute_stats(spec)
  File "/Users/nluetzge/JWST/dev_tools/specviz/specviz/specviz/plugins/statistics/statistics_widget.py", line 76, in compute_stats
    'fwhm': fwhm(spectrum),
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/width.py", line 105, in fwhm
    return computation_wrapper(_compute_fwhm, spectrum, regions)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/utils.py", line 19, in computation_wrapper
    return func(spectrum, **kwargs)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/width.py", line 173, in _compute_fwhm
    return _compute_single_fwhm(flux, spectral_axis)
  File "/Users/nluetzge/anaconda2/envs/specviz/lib/python3.6/site-packages/specutils/analysis/width.py", line 152, in _compute_single_fwhm
    r_idx = np.where(right == True)[0][0] + argmax
IndexError: index 0 is out of bounds for axis 0 with size 0
Abort trap: 6
```

**Comment by SaOgaz** [comment]
I will investigate crashes....

**Comment by robelgeda** [comment]
@SaOgaz : This should be a ticket in SpecUtils. We can just wrap it around a try-except statement for now.

**Comment by SaOgaz** [comment]
@astrofrog, this travis failure seems to be one of the glue tests. @robelgeda and I tried to figure out why it was failing, something with a copy on one of the WCS objects, but we didn't have much luck.

@hcferguson, is a glue failing test a blocker on merging into master?  (I still need to add tests and get a review before merge)

**Comment by SaOgaz** [comment]
Hey @javerbukh, tests are added, ready for review!

File: specviz/plugins/statistics/statistics_widget.py [hunk 1]
```diff
@@ -57,17 +58,49 @@ def compute_stats(spectrum):
...
-    mean = flux.mean()
-    rms = np.sqrt(flux.dot(flux) / len(flux))
-    return {'mean': mean,
-            'median': np.median(flux),
-            'stddev': flux.std(),
+
+    try:
+        cent = centroid(spectrum, region=None)  # we may want to adjust this for continuum subtraction
+    except Exception as e:
+        print("centroid err: ",e)
```
Comment (nmearl) [inline specviz/plugins/statistics/statistics_widget.py]: Can you remove this print statement? Also, could you add all the exception output as logging errors?

```python
try:
    ...
except Exception as e:
    logging.error(e)
    ...

File: specviz/plugins/statistics/statistics_widget.py [hunk 2]
```diff
@@ -143,15 +190,17 @@ def _update_stat_widgets(self, stats):
             return
         for key in stats:
             if key in self.stat_widgets:
-                text = format_float_text(stats[key])
-                self.stat_widgets[key].setText(text)
+                print(key, " ", stats[key])
```
Comment (nmearl) [inline specviz/plugins/statistics/statistics_widget.py]: Remove print statement.

File: specviz/plugins/statistics/statistics_widget.py [hunk 3]
```diff
@@ -210,6 +259,38 @@ def clear_statistics(self):
         self._clear_stat_widgets()
         self.stats = None
 
+    def _reconnect_item_signals(self):
+        if self.hub.plot_item is self._current_plot_item:
+            return
+
+        if isinstance(self._current_plot_item, PlotDataItem):
```
Comment (nmearl) [inline specviz/plugins/statistics/statistics_widget.py]: Just to clarify: is that a particular reason this was necessary? Are you expecting the plot item to be anything else?

File: specviz/plugins/statistics/statistics_widget.py [hunk 4]
```diff
@@ -222,12 +303,19 @@ def update_statistics(self):
         spectral_region = self._get_workspace_region()
 
         self._current_spectrum = spec
+        self._reconnect_item_signals()
```
Comment (nmearl) [inline specviz/plugins/statistics/statistics_widget.py]: I don't think you should be running this test every time the update statistics is called. Why not have the plugin listen for selection changes in the workspace, and update the current item reference there?

```python
# In your __init__
self.hub.workspace.current_selection_changed.connect(lambda plot_item: self.update_statistics())
```

The signal returns the currently selected item, which could be passed to the update statistics function.

File: specviz/plugins/statistics/statistics_widget.py (hunk 3 above)
Comment (robelgeda) [reply specviz/plugins/statistics/statistics_widget.py]: Nope, this is just a blanket check for `None` and anything that is not a `PlotDataItem` introduced in the future.

File: specviz/plugins/statistics/statistics_widget.py (hunk 4 above)
Comment (robelgeda) [reply specviz/plugins/statistics/statistics_widget.py]: We would still need a way to listen to unit changes. The main reason for this line is to make sure the stats widget updates when a unit changes. This is done by listening to `PlotDataItem.data_unit_changed` and `PlotDataItem.data_unit_changed`. I agree this is not the best way of doing it but there isn't a single signal for unit changes yet. Let me know if there is a better way to do this?

**Comment by robelgeda** [comment]
@astrofrog @eteq @brechmos-stsci : @SaOgaz  and I spent quite a bit of time looking at the bug she mentioned. The reason the test is failing is because we are trying to deepcopy a `Spectrum1D` object initialized with `flux` and `wcs` kwargs. I tried to deepcopy a similar `Spectrum1D` in ipython and I get the same error. When I initialize with `flux` and `spectral_axis` only, it copies without an issue. If this is not unique to me, I can make a ticket.

**Comment by nmearl** [comment]
I've put a fix for this in astropy/specutils/pull/390.

File: specviz/plugins/statistics/statistics_widget.py (hunk 2 above)
Comment (SaOgaz) [reply specviz/plugins/statistics/statistics_widget.py]: Sorry about this, I thought I had checkout a clean version of stuff, but clearly I hadn't 🤣

**Comment by javerbukh** [comment]
Pulled this branch. Loaded Specviz with the COS_FUV.fits file, then added a region. I tried to resize the region but every pixel I moved came with this output in terminal `ERROR:root:Spectrum1D currently requires the uncertainty be defined.` It made the region moving extremely laggy, not sure if anyone else got this bug.

**Comment by SaOgaz** [comment]
I've never seen the slow re-sizing problem.  The Error might be coming from one of the calculations.  @robelgeda, would you have time to look at this?

**Comment by brechmos-stsci** [comment]
@javerbukh @SaOgaz The specutils SNR requires the uncertainty be defined in the Spectrum1D object. So, if it is not defined then you will get the error.  Specviz will have to decide what to do about it. There is a method in specutils to define an uncertainty based on an RMS type error of the spectrum, so that is one option.

@robelgeda

**Comment by SaOgaz** [comment]
I'm pretty sure we decided that the no SNR calculation was acceptable, hence filling that box in with a NaN.  But, if the logging is slowing down the API, we should address that.

**Comment by nluetzge** [comment]
Hmmm I used to to use this formular alot. It doesnt need uncertainies but I’m not sure how accurate it still is. Maybe @eteq or @gderosa2004 can weigh in? http://www.stecf.org/software/ASTROsoft/DER_SNR/

**Comment by brechmos-stsci** [comment]
@nluetzge Maybe post an issue in specutils as an enhancement to add another SNR calculation method.

**Comment by hcferguson** [comment]
That algorithm looks quite handy; agreed it's best to put in SpecUtils first. The link to the newsletter article (which justifies the algorithm) in the python script didn't work, but a little googling led to this https://www.spacetelescope.org/static/archives/stecfnewsletters/pdf/hst_stecf_0042.pdf.

**Comment by nluetzge** [comment]
Ok cool. Will open an issue in specutils then.

**Pull request merged**